### PR TITLE
docs(grpcplugins): add how to develop a grpc plugins for CDS

### PR DESCRIPTION
**Pull request by bnjjj**
Signed-off-by: Benjamin Coenen <benjamin.coenen@corp.ovh.com>

1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: contrib/grpcplugins/action/examples/nodejs/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# CDS GRPC Plugin example in Node.Js
+
+Here is an example of CDS plugin working with GRPC in Node.Js. If you want to test or if you want to create a GRPC Plugin in Node.Js please be aware that when you use GRPC with Node.Js, inside `node_modules` there is a binary which is downloaded for your host OS. Moreover we wrote a SDK for NodeJs [just here](https://github.com/ovh/cds/tree/master/sdk/grpcplugin/actionplugin/nodejs) with a [README](https://github.com/ovh/cds/tree/master/sdk/grpcplugin/actionplugin/nodejs/README.md) to explain you how to use it. So if you want to build and publish the plugin for linux/amd64 you must be on a linux/amd64. There already an example of `Makefile` which is working, feel free to update it for your needs.
```
Comment (rayene) [inline contrib/grpcplugins/action/examples/nodejs/README.md]: Here is an example of a Node.Js CDS plugin using GRPC. If you want to create a GRPC Plugin in Node.Js, make sure to install the binary corresponding to your host OS inside the `node_modules` directory. Moreover we wrote a SDK for NodeJs [just here]

Comment (rayene) [inline contrib/grpcplugins/action/examples/nodejs/README.md]: There already an example of a working `Makefile`. Feel free to adapt it to your needs.

File: docs/content/workflows/pipelines/actions/plugins/write-plugin.md [hunk 2]
```diff
@@ -10,10 +10,16 @@ A step is :
 
 * a builtin action, as GitClone, etc... [read more]({{< relref "workflows/pipelines/actions/builtin/_index.md" >}})
 * a user action [read more]({{< relref "workflows/pipelines/actions/user-actions/_index.md" >}})
-* a Plugin Action
+* a plugin action
 
-A Plugin is a Golang Binary.
+A Plugin is simply an executable which expose a GRPC server corresponding to the right [proto file](https://github.com/ovh/cds/tree/master/sdk/grpcplugin/actionplugin/actionplugin.proto). You can use any languages. The CDS worker will simply query in GRPC the plugin (which is the GRPC server).
```
Comment (rayene) [inline docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: You can use the programming language of your choice. The CDS worker will simply query the GRPC server of the plugin.

File: docs/content/workflows/pipelines/actions/plugins/write-plugin.md [hunk 3]
```diff
@@ -10,10 +10,16 @@ A step is :
...
 
-Take a look at https://github.com/ovh/cds/tree/master/sdk/plugin/dummy/dummy_plugin.go
+So a CDS plugin have these requirements in order to communicate with worker:
```
Comment (rayene) [inline docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: In order to communicate with a CDS worker, a plugin MUST fill the following requirements:

File: docs/content/workflows/pipelines/actions/plugins/write-plugin.md [hunk 4]
```diff
@@ -10,10 +10,16 @@ A step is :
...
 
-Contribute on https://github.com/ovh/cds/tree/master/contrib/plugins
++ Must expose a GRPC server
```
Comment (rayene) [inline docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: Must expose -> Expose

File: docs/content/workflows/pipelines/actions/plugins/write-plugin.md [hunk 5]
```diff
@@ -10,10 +10,16 @@ A step is :
...
++ Must implement methods and messages coming from this [proto file](https://github.com/ovh/cds/tree/master/sdk/grpcplugin/actionplugin/actionplugin.proto)
```
Comment (rayene) [inline docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: Must implement -> Implement

File: docs/content/workflows/pipelines/actions/plugins/write-plugin.md [hunk 6]
```diff
@@ -10,10 +10,16 @@ A step is :
...
++ And last but not least at the launch of your plugin you can use random unix socket or random tcp port but in order to inform worker what is your address you have to display this log at the launch of your plugin `XXX is ready to accept new connection` where `XXX` is your ip address with port or your unix socket (example: `127.0.0.1:55939 is ready to accept new connection`).
```
Comment (rayene) [inline docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: - Display this log at the launch of your plugin `XXX is ready to accept new connection` where `XXX` is your ip address with port or your unix socket (example: `127.0.0.1:55939 is ready to accept new connection`). Note that your plugin can use any unix socket or tcp port as long as it informs the worker using the log line above.

Comment (rayene) [reply docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: [technical] I find this way of interaction a bit ugly :)

File: docs/content/workflows/pipelines/actions/plugins/write-plugin.md [hunk 7]
```diff
@@ -10,10 +10,16 @@ A step is :
...
+
+We give you some resources to help you develop a CDS plugin like [SDK in this directory](https://github.com/ovh/cds/tree/master/sdk/grpcplugin/actionplugin) and examples [here](https://github.com/ovh/cds/tree/master/contrib/grpcplugins/action/examples).
```
Comment (rayene) [inline docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: Rule of thumb: Never use "We".

More resources that may help you in developing a CDS plugin are available at/here ...

Comment (rayene) [reply docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: with some examples here

File: sdk/grpcplugin/actionplugin/nodejs/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+# SDK Node.Js for CDS GRPC action plugins
+
+## How to use
+
++ First you have to create a link on npm locally with this package, so in this directory you just have to execute:
```
Comment (rayene) [inline sdk/grpcplugin/actionplugin/nodejs/README.md]: First you have to create a link on npm locally with this package by typing the following command in its directory:

File: sdk/grpcplugin/actionplugin/nodejs/README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+$ npm link
+```
+
++ Then go to the directory of your plugin and launch:
+
+```
+$ npm link sdk-grpcplugin-cds-nodejs
+```
+
++ That's all you just have to write:
```
Comment (rayene) [inline sdk/grpcplugin/actionplugin/nodejs/README.md]: This is unclear.

File: sdk/grpcplugin/actionplugin/nodejs/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+function main() {
+  let client = new sdk.Client(path.join(__dirname, './nodejs.yml'), run); //Indicate the yaml file which describe your plugin
+  client.start();
+}
+
+main();
+```
+
+For a more complete example you can [go there](https://github.com/ovh/cds/tree/master/contrib/grpcplugins/action/examples/nodejs)
```
Comment (rayene) [inline sdk/grpcplugin/actionplugin/nodejs/README.md]: You can find a richer example [here]

File: docs/content/workflows/pipelines/actions/plugins/write-plugin.md (hunk 6 above)
Comment (bnjjj) [reply docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: Yes agree with you, it's not the best way. I'll work for a better way later. At the moment I try to stay ISO as before

File: contrib/grpcplugins/action/examples/nodejs/README.md [hunk 11]
```diff
@@ -0,0 +1,3 @@
+# CDS GRPC Plugin example in Node.Js
+
+Here is an example of a Node.Js CDS plugin using GRPC. If you want to create a GRPC Plugin in Node.Js, make sure to install the binary corresponding to your host OS inside the `node_modules` directory. Moreover we wrote a SDK for NodeJs [just here](https://github.com/ovh/cds/tree/master/sdk/grpcplugin/actionplugin/nodejs) with a [README](https://github.com/ovh/cds/tree/master/sdk/grpcplugin/actionplugin/nodejs/README.md) to explain you how to use it. So if you want to build and publish the plugin for linux/amd64 you must be on a linux/amd64. There already an example of a working Makefile. Feel free to adapt it to your needs.
```
Comment (fsamin) [inline contrib/grpcplugins/action/examples/nodejs/README.md]: node.JS

File: docs/content/workflows/pipelines/actions/plugins/write-plugin.md [hunk 12]
```diff
@@ -10,10 +10,16 @@ A step is :
...
-A Plugin is a Golang Binary.
+A Plugin is simply an executable which expose a GRPC server corresponding to the right [proto file](https://github.com/ovh/cds/tree/master/sdk/grpcplugin/actionplugin/actionplugin.proto). You can use the programming language of your choice. The CDS worker will simply query the GRPC server of the plugin.
 
-Take a look at https://github.com/ovh/cds/tree/master/sdk/plugin/dummy/dummy_plugin.go
+In order to communicate with a CDS worker, a plugin MUST fill the following requirements:
 
-Contribute on https://github.com/ovh/cds/tree/master/contrib/plugins
++ Expose a GRPC server
++ Implement methods and messages coming from this [proto file](https://github.com/ovh/cds/tree/master/sdk/grpcplugin/actionplugin/actionplugin.proto)
++ Display this log at the launch of your plugin XXX is ready to accept new connection where XXX is your ip address with port or your unix socket (example: 127.0.0.1:55939 is ready to accept new connection). Note that your plugin can use any unix socket or tcp port as long as it informs the worker using the log line above.
```
Comment (fsamin) [inline docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: s/log/message/

Comment (fsamin) [reply docs/content/workflows/pipelines/actions/plugins/write-plugin.md]: with an example and talk about unix socket

**Pull request merged**